Update for bullseye debian-mail-overlay

SaraSmiseth opened the pull request:
## Description

This PR is for the updated bullseye debian-mail-overlay which is here https://github.com/mailserver2/debian-mail-overlay/pull/11. Once that is merged this PR will fix the failing tests and do other necessary changes.

I fixed a few tests that were broken, because logging messages changed. There are still 2 tests left that do not run successfully. I'm not sure why they fail at the moment.

The failing tests are:

https://github.com/mailserver2/mailserver/blob/ab9cac04334ae9397045fbb7a06a864fce1ec3c6/test/tests.bats#L1462
https://github.com/mailserver2/mailserver/blob/ab9cac04334ae9397045fbb7a06a864fce1ec3c6/test/tests.bats#L1468

### Fixes

- Fixes #30.
- Fix traefik acme cert works correctly tests
- Fix Eicar-Test-Signature FOUND test

## Type of change

- [X] Update base image
- [X] Fix broken tests

## Status

- [ ] Ready
- [X] In development
- [ ] Hold

## TODO List

- [ ] Fix remaining failing Tests

## How has this been tested?

I have updated my server to use an image based on the updated debian-mail-overlay for about 2 weeks.

sknight80 reviewed (approved):
So far, looks good to me.

SaraSmiseth commented:
The 2 tests mentioned above still fail. I will look into them when I have time at the weekend.
It looks like Travis does not run.
I found an error message [here](https://travis-ci.com/github/mailserver2/mailserver/requests ): Could not authorize build request for mailserver2/mailserver.

AndrewSav commented:
> It looks like Travis does not run.
> I found an error message [here](https://travis-ci.com/github/mailserver2/mailserver/requests): Could not authorize build request for mailserver2/mailserver.

I think I [fixed this](https://github.com/mailserver2/mailserver/pull/38) by [selecting the free plan](https://travis-ci.community/t/could-not-authorize-build-request/11724). However my build fails, due to EICAR, which I believe you already fixed?

Probably no worth fixing my PR here, so I will just close it, but feel free to pull the update of the travis link (unrelated) to your PR from mine as well.

SaraSmiseth commented:
The reason for both `checking dovecot: piped ham message with sieve` tests failing is that the dovecot logging for sieve has changed. If I run the container with `DEBUG_MODE=true` the messages that the tests are expecting are there. It seems that they were changed from INFO to DEBUG level.
I have to change the tests so that they check for the messages in a container that is started with `DEBUG_MODE=true` . I'm not exactly sure how. Maybe add a new container just for these 2 tests? I will have to check if that's possible and we don't run into problems having too much test containers running at the same time.

AndrewSav commented:
Sorry, do not have anything useful to contribute, but everything you say makes sense to me. A separate container should work, but at the same time we already have quite a lot of them and they are quire heavy, so the concern about a new one is founded.

SaraSmiseth commented:
I have played around with github actions [here](https://github.com/SaraSmiseth/mailserver/pull/1) and got it working.

The problem I had was that random rspamd tests failed because the runners do not have enough RAM. I got it working by splitting tests and running the new container and the tests for the new container after everything else. See this ugly [commit](https://github.com/SaraSmiseth/mailserver/pull/1/commits/be85ae0420b8634744fa6f80d3156894b119e9f6).

I don't like that workaround. What do you think?

AndrewSav commented:
@SaraSmiseth what worries you? If we do not have enough memory to run them all at the same time we have to run them in stages, and that's what you did. May be to get rid of the mental sense of ugliness we just need to accept that this is a multistage tests which require different set of containers running at each stage. Would that thinking help? Or is there anything else particularly ugly that does not sit well?

SaraSmiseth commented:
> May be to get rid of the mental sense of ugliness we just need to accept that this is a multistage tests which require different set of containers running at each stage. Would that thinking help?

Yes thanks.

I added a target for each mailserver-test-container to the Makefile.
This allows to run for example only ldap tests with `make ldap`.
Running only `make` will execute all tests.
I split the test.bats file into multiple bats-test-files.

I have also added  a github actions workflow for each mailserver-test-container.
The workflows will build the docker image and then run the tests.
For example: the workflow `.github/workflows/default.yml` will run
`make default` to run tests for the default container.
These workflows run only for pull requests into master.

AndrewSav commented:
> This should be all. Should we merge it?

How urgent do you feel? I can do my best to try and give at a spin next weekend. If you are fairly confident that this works, then by all means go ahead.

AndrewSav commented:
Also can we then remove travis or something? (I'm referring to the "All check have failed" message) Because it does not look like it provides any value at this stage.

SaraSmiseth commented:
> How urgent do you feel? I can do my best to try and give at a spin next weekend. If you are fairly confident that this works, then by all means go ahead.

It's not that urgent. Take your time.

> Also can we then remove travis or something? (I'm referring to the "All check have failed" message) Because it does not look like it provides any value at this stage.

Yes I'll remove it.

sknight80 commented:
Sorry, I am a little bit late for the party.  How can I help?

AndrewSav commented:
@sknight80 by testing that the changes here are working fine for you. Thanks.

AndrewSav commented:
I keep getting 

```
✗ checking postfix: milter-reject - clamav virus found
   (from function `assert_success' in file test/test_helper/bats-assert/src/assert.bash, line 114,
    in test file test/default.bats, line 491)
     `assert_success' failed

   -- command failed --
   status : 1
   output :
   --
```

`docker exec mailserver_default cat /var/log/mail.log | grep EIC` yields:

```
2021-09-18T02:00:25.064261+00:00 mail clamd[1770]: instream(127.0.0.1@35030): Win.Test.EICAR_HDB-1(697ef5516dc83129c71542c60777da9c:790) FOUND
2021-09-18T02:00:25.330080+00:00 mail postfix/cleanup[2179]: 087FE810: milter-reject: END-OF-MESSAGE from localhost[127.0.0.1]: 5.7.1 clamav: virus found: "Win.Test.EICAR_HDB-1"; from=<virus@gmail.com> to=<john.doe@domain.tld> proto=SMTP helo=<mx.gmail.com>
```

Are they keeping changing virus found messages?

AndrewSav reviewed (changes requested):
There are three things:
- Virus detection - not sure but I'm getting inconsistent "virus found" message, the virus name flip between what's currently in the code and what I commented with
- On my machine a lot of things does not have time to start, so I needed additional delays 
-  Ideally instead of `sleep` we should _wait for the ready conditions_ but this is probably out of scope of this PR

SaraSmiseth commented on Makefile:
```diff
@@ -283,43 +372,68 @@ fixtures:
+	# Wait for clamav unofficial sigs database update (sieve)
+	docker exec mailserver_sieve /bin/sh -c "while [ -f /var/lib/clamav-unofficial-sigs/pid/clamav-unofficial-sigs.pid ] ; do sleep 1 ; done"
+	# Wait for clamav load databases (sieve)
+	docker exec mailserver_sieve /bin/sh -c "while ! echo PING | nc -z 0.0.0.0 3310 ; do sleep 1 ; done"
```
Added.

SaraSmiseth commented on Makefile:
```diff
@@ -261,19 +349,20 @@ init:
 	# Wait for clamav unofficial sigs database update (default)
 	docker exec mailserver_default /bin/sh -c "while [ -f /var/lib/clamav-unofficial-sigs/pid/clamav-unofficial-sigs.pid ] ; do sleep 1 ; done"
 	# Wait for clamav load databases (default)
 	docker exec mailserver_default /bin/sh -c "while ! echo PING | nc -z 0.0.0.0 3310 ; do sleep 1 ; done"
```
Added.

SaraSmiseth commented on Makefile:
```diff
@@ -161,6 +112,32 @@ init:
+	# Wait for clamav unofficial sigs database update (ldap)
+	docker exec mailserver_ldap /bin/sh -c "while [ -f /var/lib/clamav-unofficial-sigs/pid/clamav-unofficial-sigs.pid ] ; do sleep 1 ; done"
+	# Wait for clamav load databases (ldap)
+	docker exec mailserver_ldap /bin/sh -c "while ! echo PING | nc -z 0.0.0.0 3310 ; do sleep 1 ; done"
```
Added.

SaraSmiseth commented on Makefile:
```diff
@@ -29,91 +52,19 @@ init:
-		-t $(NAME)
+		-t postgres:13-alpine
 
+init_ldap: init_openldap init_redis
```
Added rm to every init.

SaraSmiseth commented on test/default.bats:
```diff
@@ -0,0 +1,850 @@
+}
+
+@test "checking postfix: milter-reject - clamav virus found" {
+  run docker exec mailserver_default grep -i 'milter-reject.*virus found: "{HEX}EICAR.TEST.3.UNOFFICIAL"; from=<virus@gmail.com>' /var/log/mail.log
```
Yes these messages keep changing. Right now its always the same message for me. We can probably change the grep so that it find both versions.

AndrewSav replied:
Here is what I did here:
`run docker exec mailserver_default grep -i 'milter-reject.*virus found: ".*EICAR.*"; from=<virus@gmail.com>' /var/log/mail.log`

AndrewSav commented on test/default.bats:
```diff
@@ -0,0 +1,850 @@
+}
+
+@test "checking clamav: Eicar-Test-Signature FOUND" {
+  run docker exec mailserver_default grep -i 'EICAR\.TEST\.3\.UNOFFICIAL.*FOUND' /var/log/mail.log
```
Here is what I did here: `run docker exec mailserver_default grep -i 'EICAR.*FOUND' /var/log/mail.log`

AndrewSav commented:
@SaraSmiseth what about `sleep`s? I think we should add them to make the test work on a bigger range of machines until such time someone has time to substitute them with proper waits for readiness?

SaraSmiseth commented:
I have added the sleeps and updated EICAR tests. I agree that the sleeps should be changed later.

SaraSmiseth commented on Makefile:
```diff
@@ -1,23 +1,46 @@
+	docker run \
+		-d \
+		--name redis \
+		-t redis:6.2-alpine
```
Done.

AndrewSav reviewed (approved):
Ok, I think we are good. I also ran an instance of it in a lab for last couple of days it seems fine.

AndrewSav commented:
@SaraSmiseth thank you for your work on this upgrade. A bit of unrelated question, I know you are not using rainloop, what mail client are you using with the mailserver?

SaraSmiseth commented:
> @SaraSmiseth thank you for your work on this upgrade. A bit of unrelated question, I know you are not using rainloop, what mail client are you using with the mailserver?

Thank you for the review. I use thunderbird and no webmail at all.
